src/tests/testClpMap.cc

```diff
@@ -0,0 +1,203 @@
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyD.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyD.entries());
+}
+
+void
+testClpMap::testSetMemLimit()
+{
+    TestMap m(2048);
+    // overflow the map with entries to make sure it has lots of entries to purge below
+    fillMapWithElements(m, 10);
+    const auto testEntriesBefore = m.entries();
+    CPPUNIT_ASSERT(testEntriesBefore > 0);
+
+    m.setMemLimit(m.memoryUsed() / 2);
+    const auto entriesAfterPurge = m.entries();
+    CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
+
+    m.setMemLimit(m.memLimit() * 2);
+    // overflow the map with entries again to make sure it can grow after purging
+    fillMapWithElements(m, 10);
+    CPPUNIT_ASSERT(entriesAfterPurge < m.entries());
+}
+
+void
+testClpMap::testTtlExpiration()
+{
+    TestMap m(2048);
+    m.add(std::to_string(1), 1, 10);
+    CPPUNIT_ASSERT(m.get("1"));
+    squid_curtime += 100;
+    // "1" should have expired
+    CPPUNIT_ASSERT(!m.get("1"));
+}
+
+void
+testClpMap::testReplaceEntryWithShorterTtl()
+{
+    TestMap m(2048);
+    addSequenceOfElementsToMap(m, 1, 0, 100);
+    CPPUNIT_ASSERT(m.get("0")); // successfully added one element
+    squid_curtime += 20;
+    CPPUNIT_ASSERT(m.get("0")); // hasn't expired yet
+    squid_curtime += 100;
+    CPPUNIT_ASSERT(!m.get("0")); // has expired
+
+    addSequenceOfElementsToMap(m, 1, 0, 100);
+    addSequenceOfElementsToMap(m, 1, 0, 10); // replaced element with same but shorter ttl
+    squid_curtime += 20;
+    CPPUNIT_ASSERT(!m.get("0")); // should have expired
+}
+
+void
+testClpMap::testEntriesWithZeroTtl()
+{
+    TestMap m(2048);
+    addSequenceOfElementsToMap(m, 1, 0, 0);
+    CPPUNIT_ASSERT(m.get("0")); // we get something
+    squid_curtime += 1;
+    CPPUNIT_ASSERT(!m.get("0")); // expired, we get nothing
+}
+
+void
+testClpMap::testEntriesWithNegativeTtl()
+{
+    TestMap m(2048);
+    CPPUNIT_ASSERT(!m.add("0", 0, -1)); // failure on insertion
+    CPPUNIT_ASSERT(!m.get("0"));  // we get nothing
+    CPPUNIT_ASSERT(m.add("0", 1, 0));
+    CPPUNIT_ASSERT(m.get("0"));  // we get something
+    CPPUNIT_ASSERT(!m.add("0", 2, -1));  // failure on insertion
+    CPPUNIT_ASSERT(!m.get("0"));  // we get nothing
+}
+
+void
+testClpMap::testPurgeIsLRU()
+{
+    TestMap m(2048);
+    // TODO: once we have ClpMap iterators we can inspect the contents instead
+    for (int j = 0; j < 10; ++j)
+        CPPUNIT_ASSERT(m.add(std::to_string(j), j, 10));
+    // this makes it the LRU
+    CPPUNIT_ASSERT(m.get("0"));
+    // now overflow the map, checking "0" each time we add an element
+    for (int j = 100; j < 1000; ++j) {
+        CPPUNIT_ASSERT(m.add(std::to_string(j), j, 10));
+        CPPUNIT_ASSERT(m.get("0"));
+    }
+    CPPUNIT_ASSERT(m.get("0"));
+    CPPUNIT_ASSERT(!m.get("1")); // these should have been aged out
+    CPPUNIT_ASSERT(!m.get("2"));
+    CPPUNIT_ASSERT(!m.get("3"));
+    CPPUNIT_ASSERT(!m.get("4"));
+}
```

Comment:
Here, I would overflow the map again (using our fillMapWithElements()) and check that "0" is now gone. This would indirectly confirm that there is nothing special about that "0" element except our get() accesses to it in the loop above

```suggestion

    fillMapWithElements(m, 10);
    CPPUNIT_ASSERT(!m.get("0")); // removable when not recently used
}
```